### DISCUSSION TITLE
Kernel mode eBPF program loader API

### DISCUSSION THREAD

**Discussion by nigriMSFT** [microsoft/ebpf-for-windows#2975] [Ideas]
Is there a plan to expose eBPF program loader APIs (bpf APIs) to kernel mode, so that a driver can load programs itself?

**Comment by shankarseal** [comment]
Thanks @nigriMSFT. There are no current plans. However if there is a scenario, please provide details and we will consider it.

**Comment by saxena-anurag** [comment]
One possible solution is to add support for static initialization of native module service. (Today a service entry is dynamically created by eBPFAPI and that service is then marked delete pending once the program is loaded -- all this is done to streamline user mode app workflow; user mode app should be able to use libbpf APIs even for native module, and details about service creation / deletion should be abstracted out from it). With the support of static service entries, it still requires a pre-setup step where a user mode service is created once from user mode, and _then_ that service can be used to load programs from kernel.

So, in the new flow, the caller will now be responsible for the lifecycle of the created service. We may also need to add a new _eBPF_ API which takes a service as input (instead of file name).
